lib/config/config.go

```diff
@@ -362,14 +345,92 @@ func (cfg *Configuration) clean() error {
 
 	// The list of ignored devices should not contain any devices that have
 	// been manually added to the config.
-	newIgnoredDevices := []protocol.DeviceID{}
+	var newIgnoredDevices []ObservedDevice
+	ignoredDevices := make(map[protocol.DeviceID]bool)
 	for _, dev := range cfg.IgnoredDevices {
-		if !existingDevices[dev] {
+		if !existingDevices[dev.ID] {
+			ignoredDevices[dev.ID] = true
 			newIgnoredDevices = append(newIgnoredDevices, dev)
 		}
 	}
 	cfg.IgnoredDevices = newIgnoredDevices
 
+	// The list of pending devices should not contain devices that were added manually, nor should it contain
+	// ignored devices.
+
+	// Sort by time, so that in case of duplicates latest "time" is used.
+	sort.Slice(cfg.PendingDevices, func(i, j int) bool {
+		return cfg.PendingDevices[i].Time.Before(cfg.PendingDevices[j].Time)
+	})
+
+	var newPendingDevices []ObservedDevice
+nextPendingDevice:
+	for _, pendingDevice := range cfg.PendingDevices {
+		if !existingDevices[pendingDevice.ID] && !ignoredDevices[pendingDevice.ID] {
+			// Deduplicate
+			for _, existingPendingDevice := range newPendingDevices {
+				if existingPendingDevice.ID == pendingDevice.ID {
+					continue nextPendingDevice
+				}
+			}
+			newPendingDevices = append(newPendingDevices, pendingDevice)
+		}
+	}
```

Comment:
Building the map, then converting the map back to an array I suspect is going to be the same amount of code.

Reply:
Since we're bikeshedding this piece of code, here's what I would do.

```
var newPendingDevices []ObservedDevice
pendingIndexes := map[DeviceID]int // device ID to index in newPendingDevices
for (blah in old pending devices) {
  if ignoredPendingDevices[blah] {
    // skip ignored devices
    continue
  }
  idx, ok := pendingIndexes[blah]
  if !ok {
    // entry is not in newPendingDevices
    pendingIndexes[blah] = len(newPendingDevices)
    newPendingDevices = append(...)
  } else if newPendingDevices[idx].time.before(the new time) {
    // we should replace with the newer entry
    newPendingDevcies[idx] = blah // overwrite
  }
}
```

I suspect something similar could be used in dedup steps below to do things in one pass instead of nested loops, but I haven't thought it through